### PR TITLE
writr-rs: byte-exact Rust markdown engine (native + wasm + browser) with multi-core rendering

### PR DESCRIPTION
**Pull request by jaredwray**
**Please check if the PR fulfills these requirements**
- [x] Followed the [Contributing](../blob/main/CONTRIBUTING.md) and [Code of Conduct](../blob/main/CODE_OF_CONDUCT.md) guidelines.
- [x] Tests for the changes have been added (for bug fixes/features) with 100% code coverage. *(Rust workspace: ~98% lines with every uncovered line documented as unreachable-by-construction in `writr-rs/COVERAGE.md`; `writr-node` is host-bound and covered end-to-end by the harness. The JS suite's 100% gate is untouched and green.)*

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature — `writr-rs`, a Rust implementation of writr's markdown engine that reproduces the JS unified/remark/rehype pipeline **byte-for-byte**, shipped as a native Node addon, a WebAssembly module, and a browser ESM entry point. The JS engine and public API are untouched (`src/**` has zero changes); the only JS-side changes are the harness adapter registration, build scripts, and a new benchmark.

## What's inside

- **`writr-rs/` Cargo workspace** (~15K lines of Rust + generated tables):
  - `writr-core` — frontmatter, markdown-rs parsing, mdast transforms (GFM alerts, toc, emoji, autolink-literal fixups), hast pipeline (raw HTML via html5ever replay, slugs, highlight, KaTeX), and a hast-util-to-html@9-exact serializer.
  - `writr-hljs` — a port of highlight.js 11.11.1's engine (compiler, resumable multi-regex, keyword classifier, emitter) plus 36 grammars generated from the real hljs package.
  - `writr-katex` — the real `katex.min.js` 0.16.45 embedded in QuickJS for byte-exact math.
  - `writr-conformance` — renders the repo's golden harness from Rust and byte-diffs.
  - `writr-node` — napi-rs v3 bindings: `render`, `renderAsync`, `renderBatch` (multi-core), `renderBatchAsync`, `renderBatchBuffer` / `renderBatchBufferAsync` (packed bytes-in/bytes-out batches), `validate`, `renderToMdast`, `engineVersion`, plus wasm + browser loaders.
  - `vendor/markdown` — vendored markdown-rs 1.0.0 with parity + performance patches, all documented in `VENDORED.md`.

## Byte-exact parity (the migration safety net)

| Suite | Scale | Result |
| ----- | ----- | ------ |
| Golden harness, all 7 profiles, in-Rust | 2,041 goldens | byte-exact, **allowlist empty** |
| Same suite through the napi addon (`HARNESS_ENGINE=writr-rust`) | 2,041 | byte-exact |
| Same suite through the wasm build (`WRITR_RS_FORCE_WASM=1`) | 2,041 | byte-exact |
| highlight.js oracle fixtures (generated from real hljs) | 2,111 | byte-exact |
| Headless-Chromium browser smoke (`browser.js`, no COOP/COEP needed) | 11 docs | byte-identical to native |

`pnpm test` (the JS suite) is untouched and green. Every performance commit below was individually gated on these suites staying byte-exact.

## Performance

Benchmarked with `benchmark/benchmark-rust.ts` (tinybench, same corpus as the existing benchmarks; 4 shared vCPUs — gaps widen with real cores):

**Whole-corpus throughput** (101 docs/call):

| Engine | minimal profile | default profile (full pipeline) |
| ------ | ---- | ---- |
| **writr-rs `renderBatchBuffer` (bytes in/out, all cores)** | **~34,000 docs/s 🥇** | — |
| **writr-rs `renderBatch` (all cores)** | **~33,000 docs/s** | **~9,700 docs/s** |
| markdown-it (single-thread loop) | ~22,000 docs/s | not comparable (no highlight/math/slug/toc) |
| marked (single-thread loop) | ~15,000 docs/s | not comparable |
| writr JS | ~1,700 docs/s | ~690 docs/s |

**Single-document latency**: writr-rs sync is ~5–6.5× writr-JS per call on every profile, uncached, with a synchronous API (~4,000 ops/s on the full default pipeline vs ~565 for writr-JS). With writr's cache on top, repeat renders stay in the millions of ops/s.

The dedicated optimization pass (last 8 commits) cut the default profile another ~13%, the highlight-heavy profiles ~15%, and the wasm build ~9% on top of the original engine, via: streaming serializer (attributes + entity escaping written straight into a capacity-seeded buffer), a copy-free hljs hot loop (an O(n²) remainder copy eliminated, span-based keyword segmentation), 40-byte `u32` parser events (halving event memory traffic; vendored parser now ~1.6× upstream markdown-rs 1.0.0), `wasm-opt -O3` + SIMD128 for the wasm artifact (−14% module size), the packed `renderBatchBuffer` API (removes the serial marshalling fraction as cores grow), and an opt-in PGO build (`pnpm build:rs:pgo`, +2–5%).

## Coverage & CI

- `cargo llvm-cov` line-coverage gate for the workspace (97.75% lines at introduction, CI gate at 97; `writr-node` requires a Node host and is covered end-to-end by the harness runs). Every uncovered line is documented as unreachable-by-construction in `writr-rs/COVERAGE.md`. The suites include 248 byte-exact oracle cases captured from the real JS engine plus ~130 new targeted unit/integration tests.
- Oracle testing also surfaced a short list of edge-case behavior differences vs the JS engine on inputs *outside* the golden corpus (template-content slugs, trailing-empty comma lists, CDATA in foreign content, …) — documented in `writr-rs/KNOWN-DIVERGENCES.md` as follow-up candidates rather than silently shipped.
- New `.github/workflows/writr-rs.yml`: fmt + clippy `-D warnings` + tests + coverage gate, codegen-freshness check (regenerates all tables from the pinned npm packages and diffs), harness parity (native, wasm, browser smoke, JS suite), and a native build matrix.
- Criterion stage benches (`cargo bench -p writr-core`).

## How to try it

```sh
pnpm build:rs && HARNESS_ENGINE=writr-rust pnpm test:harness
pnpm build:rs:wasm && HARNESS_ENGINE=writr-rust WRITR_RS_FORCE_WASM=1 pnpm test:harness
npx tsx benchmark/benchmark-rust.ts
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01U2S9pW3ucDuNmi2DNu1WiJ

---
_Generated by [Claude Code](https://claude.ai/code/session_01U2S9pW3ucDuNmi2DNu1WiJ)_

### CODE REVIEW

File: writr-rs/crates/writr-core/src/mdast_util/toc.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+				return;
+			}
+
+			if let Some(opening_depth) = search.opening_depth {
+				if search.end_index.is_none() && heading.depth <= opening_depth {
+					search.end_index = Some(position);
+				}
+			}
+
+			if search.end_index.is_some() {
```
[the remark this answers is not in the document]
Comment (jaredwray) [reply writr-rs/crates/writr-core/src/mdast_util/toc.rs:97]: Checked against the JS engine (the parity contract here): for `# Contents` followed only by deeper headings, writr's remark-toc configuration also generates no TOC list — both engines produce byte-identical output (`<h1 id="contents">…</h1><h2 …>…` with no inserted list). Example oracle run:

```
input: "# Contents\n\n## Alpha\n\ntext\n\n### Beta\n\n## Gamma\n"  (toc: true, slug: true)
writr JS  => <h1 id="contents">Contents</h1>\n<h2 id="alpha">Alpha</h2>\n<p>text</p>\n<h3 id="beta">Beta</h3>\n<h2 id="gamma">Gamma</h2>
writr-rs  => identical
```

So this port matches the reference behavior; no change made.

---
_Generated by [Claude Code](https://claude.ai/code)_
